### PR TITLE
Reenable support for fallback formatter to join (#2040)

### PR DESCRIPTION
**Pull request by kamibo**
I agree that my contributions are licensed under the {fmt} license, and agree to future changes to the licensing.

### CODE REVIEW

File: include/fmt/ostream.h [hunk 1]
```diff
@@ -147,6 +147,19 @@ struct fallback_formatter<T, Char, enable_if_t<is_streamable<T, Char>::value>>
     return std::copy(buffer.begin(), buffer.end(), ctx.out());
   }
 };
+
+template <typename It, typename Sentinel, typename Char>
+struct fallback_formatter<arg_join<It, Sentinel, Char>, Char>
```
Comment (vitaut) [inline include/fmt/ostream.h:152]: `fallback_formatter` is not an extension point so it shouldn't be specialize for `arg_join`. Instead make `formatter` specialization enabled if there is a fallback formatter for the element type as it is done for ranges.

**Review by vitaut — CHANGES_REQUESTED** [review]
Thanks for the PR.

File: include/fmt/ostream.h (hunk 1 above)
Comment (kamibo) [reply include/fmt/ostream.h]: Ok hope it is better now.

File: include/fmt/format.h [hunk 2]
```diff
@@ -3720,20 +3720,32 @@ struct arg_join : detail::view {
 };
 
 template <typename It, typename Sentinel, typename Char>
-struct formatter<arg_join<It, Sentinel, Char>, Char>
-    : formatter<typename std::iterator_traits<It>::value_type, Char> {
+struct formatter<arg_join<It, Sentinel, Char>, Char> {
+  using value_type = typename std::iterator_traits<It>::value_type;
+  using context_type = basic_format_context<arg_join<It, Sentinel, Char>, Char>;
```
Comment (vitaut) [inline include/fmt/format.h:3725]: This is not needed.

File: include/fmt/format.h [hunk 3]
```diff
@@ -3720,20 +3720,32 @@ struct arg_join : detail::view {
...
+  using formatter_type =
+      conditional_t<has_formatter<value_type, format_context>::value,
+                    typename context_type::template formatter_type<value_type>,
```
Comment (vitaut) [inline include/fmt/format.h:3728]: Just use `formatter<value_type, Char` here.

File: include/fmt/format.h [hunk 4]
```diff
@@ -3720,20 +3720,32 @@ struct arg_join : detail::view {
...
+                    detail::fallback_formatter<value_type, Char>>;
+
+  formatter_type formatting;
```
Comment (vitaut) [inline include/fmt/format.h:3731]: These should be private.

**Review by vitaut — CHANGES_REQUESTED** [review]

File: include/fmt/format.h [hunk 5]
```diff
@@ -3720,24 +3720,36 @@ struct arg_join : detail::view {
...
 
 template <typename It, typename Sentinel, typename Char>
-struct formatter<arg_join<It, Sentinel, Char>, Char>
-    : formatter<typename std::iterator_traits<It>::value_type, Char> {
+struct formatter<arg_join<It, Sentinel, Char>, Char> {
+  using value_type = typename std::iterator_traits<It>::value_type;
+  using formatter_type =
+      conditional_t<has_formatter<value_type, format_context>::value,
+                    formatter<value_type, Char>,
+                    detail::fallback_formatter<value_type, Char>>;
```
Comment (vitaut) [inline include/fmt/format.h:3728]: These should be private.

File: include/fmt/format.h [hunk 6]
```diff
@@ -3720,24 +3720,36 @@ struct arg_join : detail::view {
...
         ctx.advance_to(out);
-        out = base::format(*it++, ctx);
+        out = formatting.format(*it++, ctx);
       }
     }
     return out;
   }
+
+ private:
+  formatter_type formatting;
```
Comment (vitaut) [inline include/fmt/format.h:3752]: nit: please move the private section to the beginning for consistency with other code

**Review by vitaut — CHANGES_REQUESTED** [review]

File: include/fmt/format.h (hunk 6 above)
Comment (kamibo) [reply include/fmt/format.h]: sorry, I have thought about this, no idea why I didn't do it :sweat:  
Fixed now

File: include/fmt/format.h [hunk 7]
```diff
@@ -3720,20 +3720,33 @@ struct arg_join : detail::view {
...
-    : formatter<typename std::iterator_traits<It>::value_type, Char> {
+struct formatter<arg_join<It, Sentinel, Char>, Char> {
+ private:
+  using value_type = typename std::iterator_traits<It>::value_type;
+  using formatter_type =
+      conditional_t<has_formatter<value_type, format_context>::value,
+                    formatter<value_type, Char>,
+                    detail::fallback_formatter<value_type, Char>>;
+
+  formatter_type formatting;
```
Comment (vitaut) [inline include/fmt/format.h:3731]: nit: `formatting` -> `value_formatter_`

**Review by vitaut — CHANGES_REQUESTED** [review]
Looks good, just one more comment.

File: include/fmt/format.h (hunk 7 above)
Comment (kamibo) [reply include/fmt/format.h]: Agree. It is definitely better :smile:

**Comment by vitaut** [comment]
Merged, thanks.

**Comment by kamibo** [comment]
Thanks for your time and what you have done so far. It is really great!

**Pull request merged**